File: backend/src/main/java/com/gwangmeu/shared/api/ApiResponse.java
package com.gwangmeu.shared.api;

import com.fasterxml.jackson.annotation.JsonInclude;
import org.springframework.data.domain.Page;

import java.time.LocalDateTime;

/**
 * Wrapper uniforme pour toutes les reponses API GWANG MEU.
 * Les erreurs retournent toujours ApiResponse<Void>.
 * Les listes paginées retournent ApiResponse contenant un Page<T>.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiResponse<T>(
        boolean success,
        T data,
        String message,
        int status,
        LocalDateTime timestamp
) {

    public static <T> ApiResponse<T> ok(T data) {
        return new ApiResponse<>(true, data, "OK", 200, LocalDateTime.now());
    }

    public static <T> ApiResponse<T> ok(T data, String message) {
        return new ApiResponse<>(true, data, message, 200, LocalDateTime.now());
    }

    public static <T> ApiResponse<T> created(T data) {
        return new ApiResponse<>(true, data, "Created", 201, LocalDateTime.now());
    }

    public static ApiResponse<Void> noContent() {
        return new ApiResponse<>(true, null, "No Content", 204, LocalDateTime.now());
    }

    public static ApiResponse<Void> error(String message, int status) {
        return new ApiResponse<>(false, null, message, status, LocalDateTime.now());
    }

    public static <T> ApiResponse<PageData<T>> paginated(Page<T> page) {
        PageData<T> pageData = new PageData<>(
                page.getContent(),
                page.getNumber(),
                page.getSize(),
                page.getTotalElements(),
                page.getTotalPages(),
                page.isFirst(),
                page.isLast()
        );
        return new ApiResponse<>(true, pageData, "OK", 200, LocalDateTime.now());
    }

    /**
     * Wrapper pour les reponses paginées.
     */
    public record PageData<T>(
            java.util.List<T> content,
            int page,
            int size,
            long totalElements,
            int totalPages,
            boolean first,
            boolean last
    ) {}
}

File: frontend/lib/shared/models/api_response.dart
/// Mapping du wrapper ApiResponse<T> du backend Spring Boot.
class ApiResponse<T> {
  final bool success;
  final T? data;
  final String? message;
  final int status;

  const ApiResponse({
    required this.success,
    this.data,
    this.message,
    required this.status,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic) fromJsonT,
  ) {
    return ApiResponse<T>(
      success: json['success'] as bool? ?? false,
      data: json['data'] != null ? fromJsonT(json['data']) : null,
      message: json['message'] as String?,
      status: json['status'] as int? ?? 0,
    );
  }

  bool get isSuccess => success && data != null;
}

/// Réponse paginée : correspond à ApiResponse<PageData<T>> du backend.
class PageResponse<T> {
  final List<T> content;
  final int page;
  final int size;
  final int totalPages;
  final int totalElements;
  final bool first;
  final bool last;

  const PageResponse({
    required this.content,
    required this.page,
    required this.size,
    required this.totalPages,
    required this.totalElements,
    required this.first,
    required this.last,
  });

  factory PageResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) fromJsonT,
  ) {
    final items = (json['content'] as List<dynamic>? ?? [])
        .map((e) => fromJsonT(e as Map<String, dynamic>))
        .toList();
    return PageResponse<T>(
      content: items,
      page: json['page'] as int? ?? 0,
      size: json['size'] as int? ?? 20,
      totalPages: json['totalPages'] as int? ?? 0,
      totalElements: json['totalElements'] as int? ?? 0,
      first: json['first'] as bool? ?? true,
      last: json['last'] as bool? ?? true,
    );
  }
}

